### PR TITLE
Changes for PDNS GLB MONITOR

### PR DESCRIPTION
**Pull request by nimisha0092**

### CODE REVIEW

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+	sess, err := meta.(ClientSession).PrivateDnsClientSession()
+	if err != nil {
+		return err
+	}
+
+	idset := strings.Split(d.Id(), "/")
+
+	// Update PDNS GLB Monitor if attributes has any change
+
+	if d.HasChange(pdnsGlbMonitorName) ||
```
Comment (KumarGanesanIBM) [inline ibm/resource_ibm_private_dns_glb_monitor.go:369]: no need to have two check d.HasChange.

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+		updateMonitorOptions.SetTimeout(utimeout)
+
+		_, detail, err := sess.UpdateMonitor(updateMonitorOptions)
+
+		if err != nil {
+			return fmt.Errorf("Error updating pdns GLB Monitor:%s\n%s", err, detail)
+		}
+	}
+
+	if d.HasChange(pdnsGlbMonitorType) ||
```
Comment (KumarGanesanIBM) [inline ibm/resource_ibm_private_dns_glb_monitor.go:394]: keep this d.HasChange checks.. remove above one.

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+			return fmt.Errorf("Monitor path/expected_codes/expected_body/allow_insecure/method/headers is not supported in type TCP")
+		}
+	}
+	response, detail, err := sess.CreateMonitor(CreateMonitorOptions)
+	if err != nil {
+		return fmt.Errorf("Error creating pdns GLB monitor:%s\n%s", err, detail)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", instanceID, *response.ID))
+	d.Set(pdnsGlbMonitorID, *response.ID)
```
Comment (KumarGanesanIBM) [inline ibm/resource_ibm_private_dns_glb_monitor.go:283]: add this line into Read(). remove it from here.

File: website/docs/d/private_dns_glb_monitors.html.markdown [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `instance_id` - (Required, string) The resource instance id of the private DNS on which zones were created.
+* `zone_id` - (Required, string) The unique identifier of the private DNS zone.
```
Comment (KumarGanesanIBM) [inline website/docs/d/private_dns_glb_monitors.html.markdown:29]: I don't see any argument for ibm_dns_monitor datasource. is zone_id needed here?

File: website/docs/r/private_dns_glb_monitor.html.markdown [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+# ibm\_dns_glb_monitor
+
+Provides a private dns GLB monitor resource. This allows dns GLB monitor to be created,updated and deleted.
+
+## Example Usage
+
+```hcl
+
+resource "ibm_dns_glb_monitor" "test-pdns-glb-monitor-nw" {
```
Comment (KumarGanesanIBM) [inline website/docs/r/private_dns_glb_monitor.html.markdown:17]: give full example as you have it in the test case

File: website/docs/r/private_dns_glb_monitor.html.markdown [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `instance_id` - (Required, string) The id of the private DNS on which zone has to be created.
+* `name` - (Required, string) The name of the load balancer monitor.
+* `description` -  (Optional,string) Descriptive text of the load balancer monitor.
+* `type` - (Optional, string) The protocol to use for the health check. Currently supported protocols are 'HTTP','HTTPS' and 'TCP'.
```
Comment (KumarGanesanIBM) [inline website/docs/r/private_dns_glb_monitor.html.markdown:31]: if there is default value, specify it in the document as well

File: website/docs/r/private_dns_glb_monitor.html.markdown [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+* `port` - (Optional, int) Port number to connect to for the health check. Required for TCP checks. HTTP and HTTPS checks
```
Comment (KumarGanesanIBM) [inline website/docs/r/private_dns_glb_monitor.html.markdown:32]: default value here as well

File: website/docs/r/private_dns_glb_monitor.html.markdown [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+* `interval` - (Optional, int) The interval between each health check.
```
Comment (KumarGanesanIBM) [inline website/docs/r/private_dns_glb_monitor.html.markdown:33]: check if there is default value, specify it in the document as well

File: website/ibm.erb [hunk 9]
```diff
@@ -195,9 +195,12 @@
             <li<%= sidebar_current("docs-ibm-datasource-dns-resource-records") %>>
               <a href="/docs/providers/ibm/d/private_dns_resource_records.html">dns_resource_records</a>
             </li>
+            <li<%= sidebar_current("docs-ibm-datasource-dns-glb-monitors") %>>
+              <a href="/docs/providers/ibm/d/private_dns_glb_monitors.html">dns_glb_monitors</a>
+            </li>
           </ul>
         </li>
-        <li<%= sidebar_current("docs-ibm-datasource-function") %>>
+        <li<%= sidebar_current("docs-ibm-datasource-function") %>>n
```
Comment (KumarGanesanIBM) [inline website/ibm.erb:203]: remove `n` from here.  do not edit exiting lines if it is unnecessary

File: ibm/provider.go [hunk 10]
```diff
@@ -281,6 +281,7 @@ func Provider() terraform.ResourceProvider {
 			"ibm_dns_zones":              dataSourceIBMPrivateDNSZones(),
 			"ibm_dns_permitted_networks": dataSourceIBMPrivateDNSPermittedNetworks(),
 			"ibm_dns_resource_records":   dataSourceIBMPrivateDNSResourceRecords(),
+			"ibm-dns-glb-monitors":       dataSourceIBMPrivateDNSGLBMonitors(),
```
Comment (KumarGanesanIBM) [inline ibm/provider.go:284]: change `-` hyphen to `_` underscore.

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "GLB Monitor Modification date",
+			},
+		},
+	}
+}
+
+func resourceIBMPrivateDNSGLBMonitorCreate(d *schema.ResourceData, meta interface{}) error {
+	sess, err := meta.(ClientSession).PrivateDnsClientSession()
```
Comment (KumarGanesanIBM) [inline ibm/resource_ibm_private_dns_glb_monitor.go:179]: I do not see ibm/config.go file in this diff.  Can you please check and let me know where is it?

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 1 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go:368]: handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 2 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: done

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 3 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: handled

File: website/docs/d/private_dns_glb_monitors.html.markdown (hunk 4 above)
Comment (nimisha0092) [reply website/docs/d/private_dns_glb_monitors.html.markdown]: i removed Zone_id its not required

File: website/docs/r/private_dns_glb_monitor.html.markdown (hunk 5 above)
Comment (nimisha0092) [reply website/docs/r/private_dns_glb_monitor.html.markdown]: Added

File: website/docs/r/private_dns_glb_monitor.html.markdown (hunk 7 above)
Comment (nimisha0092) [reply website/docs/r/private_dns_glb_monitor.html.markdown]: Default values are added

File: ibm/provider.go (hunk 10 above)
Comment (nimisha0092) [reply ibm/provider.go]: handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 11 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go:179]: Config files changes are already merged for PDNS changes

File: ibm/data_source_ibm_private_dns_glb_monitors.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+				dnsMonitor[pdnsGlbMonitorExpectedBody] = instance.ExpectedBody
+			}
+		}
+		if *instance.Type == "HTTPS" {
+			dnsMonitor[pdnsGlbMonitorAllowInsecure] = *instance.AllowInsecure
+		}
+
+		dnsMonitors = append(dnsMonitors, dnsMonitor)
+	}
+	d.SetId(dataSourceIBMPrivateDNSGLBMonitorsID(d))
```
Comment (kavya498) [inline ibm/data_source_ibm_private_dns_glb_monitors.go:170]: Set all required attributes`instanceID ` is not set here

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Descriptive text of the load balancer monitor",
+			},
+
+			pdnsGlbMonitorType: {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "HTTP",
+				ValidateFunc: validateAllowedStringValue([]string{"HTTP", "HTTPS", "TCP"}),
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:78]: Use Latest way of validation.. Please Refer to CIS Resources

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+		d.Set(pdnsGlbMonitorMethod, response.Method)
+		d.Set(pdnsGlbMonitorPath, response.Path)
+		d.Set(pdnsGlbMonitorExpectedCodes, response.ExpectedCodes)
+		d.Set(pdnsGlbMonitorHeaders, flattenDataSourceLoadBalancerHeader(response.HeadersVar))
+		if response.ExpectedBody != nil {
+			d.Set(pdnsGlbMonitorExpectedBody, response.ExpectedBody)
+		}
+	}
+	if *response.Type == "HTTPS" {
+		if response.AllowInsecure != nil {
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:339]: Cant this if be moved to previous if?

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+	var monitortype string
+	var monitorport int64
+
+	if monitordescription, ok := d.GetOk(pdnsGlbMonitorDescription); ok {
+		CreateMonitorOptions.SetDescription(monitordescription.(string))
+	}
+	if Mtype, ok := d.GetOk(pdnsGlbMonitorType); ok {
+		monitortype = Mtype.(string)
+	} else {
+		monitortype = "HTTP"
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:199]: Schema already defaults to `HTTP`.. Why is the monitor type made to `HTTP` any specific reason?

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+	}
+
+	if monitortype == "HTTP" {
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:202]:
```
if Mport, ok := d.GetOk(pdnsGlbMonitorPort); ok {
 			monitorport = int64(Mport.(int))
 		}
```
This check would be enough..
Lets not hardcode values for `monitorport ` here.. 
API generally handles it..

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+			pdnsGlbMonitorPath: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The endpoint path to health check against",
+			},
+
+			pdnsGlbMonitorHeaders: {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:125]: Description is missing for this attribute

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 14 above)
Comment (kavya498) [reply ibm/resource_ibm_private_dns_glb_monitor.go:339]: I believe In the case of tcp it ll be `nil`. So you need not check for types..
Just have a check for nil..
If it is not nil then set else ignore..

File: website/docs/r/private_dns_glb_monitor.html.markdown [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+* `type` - (Optional, string) The protocol to use for the health check. Currently supported protocols are 'HTTP','HTTPS' and 'TCP'.Default Value:HTTP
+* `port` - (Optional, int) Port number to connect to for the health check. Required for TCP checks. HTTP and HTTPS checks should only define the port when using a non-standard port (HTTP: default 80, HTTPS: default 443).
+* `interval` - (Optional, int) The interval between each health check.
+* `retries` - (Optional, int) The number of retries to attempt in case of a timeout before marking the origin as unhealthy.
+* `timeout` - (Optional, int) The timeout (in seconds) before marking the health check as failed.
+* `method` - (Optional, string) The method to use for the health check applicable to HTTP/HTTPS based checks, the default value is 'GET'.
+* `path` - (Optional, string) The endpoint path to health check against. This parameter is only valid for HTTP and HTTPS monitors.
+* `headers` - (Optional, string) The HTTP request headers to send in the health check. It is recommended you set a Host header by default. The User-Agent header cannot be overridden. This parameter is only valid for HTTP and HTTPS monitors.
```
Comment (kavya498) [inline website/docs/r/private_dns_glb_monitor.html.markdown:53]: Please make headers is set

File: website/ibm.erb (hunk 9 above)
Comment (nimisha0092) [reply website/ibm.erb]: Done

File: ibm/data_source_ibm_private_dns_glb_monitors.go [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+		dnsMonitor[pdnsGlbMonitorInterval] = *instance.Interval
+		dnsMonitor[pdnsGlbMonitorRetries] = *instance.Retries
+		dnsMonitor[pdnsGlbMonitorTimeout] = *instance.Timeout
+		dnsMonitor[pdnsGlbMonitorDescription] = *instance.Description
+		dnsMonitor[pdnsGlbMonitorMethod] = *instance.Method
+		dnsMonitor[pdnsGlbMonitorPath] = *instance.Path
+		dnsMonitor[pdnsGlbMonitorExpectedCodes] = *instance.ExpectedCodes
+		dnsMonitor[pdnsGlbMonitorExpectedBody] = *instance.ExpectedBody
+		dnsMonitor[pdnsGlbMonitorAllowInsecure] = *instance.AllowInsecure
+		/*if instance.Description != nil {
```
Comment (kavya498) [inline ibm/data_source_ibm_private_dns_glb_monitors.go:159]: Remove all commented lines from code

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 20]
```diff
@@ -0,0 +1,464 @@
...
+				Optional:    true,
+				Description: "Descriptive text of the load balancer monitor",
+			},
+
+			pdnsGlbMonitorType: {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "HTTP",
+				ValidateFunc: InvokeValidator(ibmDNSGlbMonitor, pdnsGlbMonitorType),
+				//ValidateFunc: validateAllowedStringValue([]string{"HTTP", "HTTPS", "TCP"}),
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:80]: Remove all commented lines from code

File: ibm/resource_ibm_private_dns_glb_monitor.go [hunk 21]
```diff
@@ -0,0 +1,464 @@
...
+		return err
+	}
+	idset := strings.Split(d.Id(), "/")
+
+	getMonitorOptions := sess.NewGetMonitorOptions(idset[0], idset[1])
+	response, detail, err := sess.GetMonitor(getMonitorOptions)
+	if err != nil {
+		return fmt.Errorf("Error fetching pdns GLB Monitor:%s\n%s", err, detail)
+	}
+	d.Set("id", response.ID)
```
Comment (kavya498) [inline ibm/resource_ibm_private_dns_glb_monitor.go:302]: setting "id" is equivalent to d.SetId() in create method..
Please comment this line

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 13 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: Handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 14 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: Handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 15 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: Done

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 16 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: Done

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 17 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go:129]: handled

File: website/docs/r/private_dns_glb_monitor.html.markdown (hunk 18 above)
Comment (nimisha0092) [reply website/docs/r/private_dns_glb_monitor.html.markdown]: handled

File: ibm/data_source_ibm_private_dns_glb_monitors.go (hunk 19 above)
Comment (nimisha0092) [reply ibm/data_source_ibm_private_dns_glb_monitors.go]: handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 20 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: handled

File: ibm/resource_ibm_private_dns_glb_monitor.go (hunk 21 above)
Comment (nimisha0092) [reply ibm/resource_ibm_private_dns_glb_monitor.go]: handled

File: ibm/data_source_ibm_private_dns_permitted_network.go [hunk 22]
```diff
@@ -122,7 +122,6 @@ func dataSourceIBMPrivateDNSPermittedNetworksRead(d *schema.ResourceData, meta i
 	for _, instance := range availablePermittedNetworks.PermittedNetworks {
 		permittedNetwork := map[string]interface{}{}
 		permittedNetworkVpcCrn := map[string]interface{}{}
-		permittedNetwork[pdnsInstanceID] = instanceID
```
Comment (kavya498) [inline ibm/data_source_ibm_private_dns_permitted_network.go:125]: Why is this been removed?

**Comment by nimisha0092** [comment]
Changes merged in PR: 
https://github.com/IBM-Cloud/terraform-provider-ibm/pull/2023

**Pull request closed without merging**